### ISSUE TITLE
Expression produces a union type that is too complex to represent. ts(2590) when using ToggleGroup

### ISSUE THREAD

**Issue by GroovyUnicyclist** [tamagui/tamagui#2919]
### Current Behavior

Attempting to use the ToggleGroup component results in the following error:

`Expression produces a union type that is too complex to represent. ts(2590)`

### Expected Behavior

Using the ToggleGroup component does not produce errors

### Tamagui Version

```markdown
1.110.5 and 1.111.0
```

### Platform (Web, iOS, Android)

```markdown
All
```

### Reproduction

```markdown
https://codesandbox.io/p/devbox/stoic-glade-x6drlr?workspaceId=d5edebc7-e0e1-45ac-9de9-700b7e45de2d

Navigate to /packages/app/features/home/screen.tsx and view the ToggleGroup on line 60.

The error will be visible in the editor.
```

### System Info

```markdown
System:
    OS: Windows 11 10.0.22621
    CPU: (16) x64 AMD Ryzen 9 5900HX with Radeon Graphics        
    Memory: 32.91 GB / 63.42 GB
  Binaries:
    Node: 21.7.3 - C:\Program Files\nodejs\node.EXE
    npm: 10.8.2 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Chromium (127.0.2651.74)
    Internet Explorer: 11.0.22621.3527
  npmPackages:
    @babel/core: ^7.20.0 => 7.24.7 
    @tamagui/config: 1.111.0 => 1.111.0 
    @tamagui/lucide-icons: 1.111.0 => 1.111.0 
    @tamagui/types: 1.111.0 => 1.111.0 
    @types/react: ~18.2.45 => 18.2.79
    eslint: ^8.57.0 => 8.57.0
    eslint-config-expo: ^7.1.2 => 7.1.2
    eslint-config-prettier: ^9.1.0 => 9.1.0
    eslint-plugin-prettier: ^5.2.1 => 5.2.1
    expo: ~51.0.31 => 51.0.32
    expo-constants: ~16.0.2 => 16.0.2
    expo-linking: ~6.3.1 => 6.3.1
    expo-router: ~3.5.23 => 3.5.23
    expo-status-bar: ~1.12.1 => 1.12.1
    prettier: ^3.3.3 => 3.3.3
    react: 18.2.0 => 18.2.0
    react-dom: 18.2.0 => 18.2.0
    react-native: 0.74.5 => 0.74.5
    react-native-safe-area-context: 4.10.5 => 4.10.5
    react-native-screens: 3.31.1 => 3.31.1
    react-native-svg: 15.2.0 => 15.2.0
    react-native-web: ~0.19.10 => 0.19.12
    tamagui: 1.111.0 => 1.111.0
    typescript: ~5.3.3 => 5.3.3
```

**Comment by dohomi** [comment]
I am seeing this error as well since the most recent tamagui releases.

**Comment by natew** [comment]
Hard to believe this could be recent, we haven't touched types in a lot of releases.

If you can pin down at least a rough release where it did work that would help me confirm to check for a regression.

**Comment by MegaPanchamZ** [comment]
I just thought I'd bump this one as well; we also seem to be having issues after updating Tamagui to 1.110.5.

Previously using 1.101.3.

**Comment by natew** [comment]
@MegaPanchamZ I know it's a pain in the ass, but could you do maybe one or two binary searches to help isolate:

Like try 1.105, if that works, try 1.103? If not try 1.107? Ideally we'd get to the exact release and that should make it easy to fix, we actually do track type perforamnce and fail tests if it slips too much.

Also the entire tamagui config setup would be useful, you can try toggling various settings or simplifying themes etc. Are you using theme builder?

**Comment by natew** [comment]
Or @dohomi / @GroovyUnicyclist basically helping find the version that triggers.

**Comment by GroovyUnicyclist** [comment]
Based on this [old issue](https://github.com/tamagui/tamagui/issues/1940#issuecomment-2070758433), it's possible it's been a thing since at least February of 2023. Will spend some time trying different versions to see what I can find out.

**Comment by MegaPanchamZ** [comment]
> @MegaPanchamZ I know it's a pain in the ass, but could you do maybe one or two binary searches to help isolate:
> 
> Like try 1.105, if that works, try 1.103? If not try 1.107? Ideally we'd get to the exact release and that should make it easy to fix, we actually do track type perforamnce and fail tests if it slips too much.
> 
> Also the entire tamagui config setup would be useful, you can try toggling various settings or simplifying themes etc. Are you using theme builder?

Hey Nate,

Should time permits I will try the different permutations for you. 

We are using the config provided by the @tamagui/config? Which should be matching the version of the library (please correct me, if I am wrong).

**Comment by natew** [comment]
`@tamagui/config/v3` has much quicker types if you haven't set that up already.

**Comment by GroovyUnicyclist** [comment]
Hey @natew I think I found the [exact commit](https://github.com/tamagui/tamagui/commit/e39b3a58e30177d642fac7bb8e01d5e7d9d4271e#diff-ff7abb3ed592dcfafbccc425cba500f748ea6d70f227e5792343dc4e79b1340dR46) that caused this. I installed v1.105.1, manually added this change in the propMapper.ts file in my node_modules folder, and observed the issue being introduced.

Not really sure what is going on in this bit of code otherwise I would consider trying to make a pr myself.

**Comment by natew** [comment]
What makes you think that caused it? Because those lines shouldn’t affect the types at all so just wondering.

**Comment by GroovyUnicyclist** [comment]
Honestly I sent that message not knowing the root cause, but I just tried changing [this line](https://github.com/tamagui/tamagui/blob/master/code/core/web/src/helpers/propMapper.ts#L34) to

```
const unsetVal = styleState.conf.unset?.[key]
```

and that worked.

I think it must be some kind of quirk with how type inference works during the destructuring of styleState. I'm not sure exactly what the quirk is though, and I'm having trouble finding any resources on it. ChatGPT seems to be pretty confident that it can happen but who knows how much that is to be trusted.

**Comment by GroovyUnicyclist** [comment]
False alarm on the specific root cause, looks like whatever plugin is displaying this error in VSCode is being a bit finnicky with refreshing. It does seem like it might've been introduced in v1.105.2 though. Still not able to get the error to show up on v1.105.1.

**Comment by GroovyUnicyclist** [comment]
@natew, I put some more time into trying to narrow down a commit that may have caused the issue. I (accidentally) installed tamagui@1.105.2 and @tamagui/config@1.105.1 and discovered that the issue went away, meaning the issue might be specifically in the config package. My new (very speculative) theory is that [this commit](https://github.com/tamagui/tamagui/commit/70ac20ddcdf4999e6019963209d5fe2cea71d338) which seemingly caused a lot of config changes may have made the union types too complex. Also worth noting that using config v1, v2, or v3 made no difference in resolving the issue, and I notice this commit affected all 3 versions of the config.

Oddly enough tamagui@1.112.0 and @tamagui/config@1.105.2 work also but I don't know if I should read into that too much, since there are probably other factors being introduced.

Idk if this is enough to go off of yet, but this may be as much as I can figure out with my limited knowledge of the inner workings of tamagui and typescript packages in general. If you don't have the time to prioritize this I suppose I'll just keep diving into it as much as I'm able to.

**Comment by natew** [comment]
I appreciate you taking the time, that is super helpful. As usual Typescript continues to shock me with how insanely hard it is to make this all work. It's been a nightmare really getting everything to work and constantly running into performance issues. At times I've greatly increased type performance only to have it hit depth limits faster. And even this incredibly innocuous change seems to cause mayhem.

The settings I guess do have an effect on types, perhaps this is more incentive to get v2 out the door so we can remove those deprecated ones and simplify things... @jonsherrard  we really need to accelerate on v2.

**Comment by dohomi** [comment]
I'm not sure if that is really the reason why this error occurs, but I was able to fix it with ToggleGroup to cast the value provided to the ToggleGroup shown below. It seems to be related that the `gap` prop, when I enter the FlexStyle prop it seems to be
```
gap?: number | undefined
``` 
even though its actually a this union type of SizeToken. I fixed it with a type cast for now.
<img width="623" alt="Screenshot 2024-10-11 at 08 22 55" src="https://github.com/user-attachments/assets/d8cba652-0265-4f8c-a880-eed915cb1e22">
<img width="319" alt="Screenshot 2024-10-11 at 08 24 08" src="https://github.com/user-attachments/assets/cf885548-3a0b-4976-87f9-dacec83561ee">

**Comment by mufaddal404** [comment]
Off topic but just started react native like 3 days ago(btw im doing react for quite some time), my first day with tamagui today and ran into this issue.
but its a great nice looking library and i'll just find a work around for now. great work founders.

and yo @GroovyUnicyclist you're doing some nice digging man!!

**Comment by GroovyUnicyclist** [comment]
Thanks @mufaddal404, I tried to find some workarounds too but have been mostly unsuccessful. To be clear, the code runs completely fine. This error just shows up in VSCode and gets in the way. It seems like you can't disable this specific error or even use ts-ignore, since it happens on a component. I believe there are ways to disable ALL errors but I really don't want to do that. You could also downgrade @tamagui/config to a different version, (1.105.1 would work but 1.105.2 also worked for me somehow) but this is probably a bad idea that the Tamagui devs probably wouldn't recommend.

Also, If you do choose to do some more digging @dohomi, I can share some tips that made it easier for me to investigate. I found that for reliable results in having the file re-evaluate for errors, I had to restart the TS server (search for this command with Ctrl+ shift + P). I also had to close and re-open the file I was testing. Before I figured this out I was disabling and re-enabling my ESLint plugin, which worked sometimes, but was a bit unreliable.

Between those 3 things, that should help you (or anyone else) investigate.

**Comment by natew** [comment]
Sorry for your issues guys, I'll get back to this soon. We at least have some idea of where it breaks.

I'm also working on v2 pretty hard now, which will simplify types down quite a bit as we can remove a few settings. But I'll make sure to get some type performance improvements as well. It's likely there are tons of different ways we can eek out more.

**Comment by natew** [comment]
@mufaddal404 did you run into this on a new project? It's weird that we don't see this anywhere across all our test apps and configurations, various sites etc.

**Comment by MufaddalMustafa2211008KHICND** [comment]
@natew yup. a new expo app with with manual tamagui setup. used `@tamagui/config/v3` for config

**Comment by mufaddal404** [comment]
@natew yup. a new expo app with with manual tamagui setup. used `@tamagui/config/v3` for config

**Comment by zaynram** [comment]
I recently decided to switch to tamagui, and I ran into this issue as well after trying to test out the advanced tabs demo component in a new app created from npm create-tamagui@latest (starter-free).

I agree with @mufaddal404 - its a nice library and since it doesn't actually create any problems (past being a visual annoyance in the editor) so i am going to keep with it for the time being, and hope that it gets patched eventually.

@natew thanks for all the work you've put in to this, and good luck with v2!

**Comment by zaynram** [comment]
> It seems like you can't disable this specific error or even use ts-ignore, since it happens on a component.

I did manage to disable typescript's complaints by adding `// @ts-expect-error` to the first instance of the error.

`" When a line is preceded by a // @ts-expect-error comment, TypeScript will suppress that error from being reported; but if there’s no error, TypeScript will report that // @ts-expect-error wasn’t necessary. "` -typescript docs

It was throwing the error for the first tabs.tab found in a tabs.list before, (removing a tabs.tab would just cause the error to appear in the next one) which makes me curious as to why i didn't have to add the expect-error comment to each tabs.tab instance. 

here's some screenshots:

**w/ error:**

<img height="350" alt="ss1" src="https://github.com/user-attachments/assets/5d33a7c7-927c-4873-8051-5e9d71842ba6"/>

**w/o error | w/ `@ts-expect-error`:**

<img height="350" alt="ss2" src="https://github.com/user-attachments/assets/64322688-84ac-447e-bac4-ffa2b4ba2f4d" />
